### PR TITLE
WPF MVVM Sample App

### PR DESCRIPTION
**Pull request by dayewah**
Here's a simple sample app that I think would have been useful to me when I first came across this excellent library. Not sure how many people are doing UI apps with this but I thought I would share my strategy as a sample. Let me know if any thing is amiss. Hope it is found to be a useful addition to the library. Btw this is my first pull request on github. Hoping its all done right and I promise I won't be offended if it is rejected :)

Thanks

### CODE REVIEW

File: Samples/WpfMVVMSample/Readme.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+﻿## WPF MVVM Sample
```
[the remark this answers is not in the document]
Comment (dayewah) [reply Samples/WpfMVVMSample/Readme.md]: Cool

File: Samples/WpfMVVMSample/Readme.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+This is a simple sample showing how UnitsNet can be used to create a WPF MVVM application. I have used this strategy in a few simple engineering apps and thought I would share it as a sample to see if others might offer improvements.
+
+It performs a simple calculation allowing flexibility in the units for parameters and results. Default units for each are specified in the settings drop down.
+
+A key feature enabling this sample is the UnitToStringConverter class
```
[the remark this answers is not in the document]
Comment (dayewah) [reply Samples/WpfMVVMSample/Readme.md]: Will do.

File: Samples/WpfMVVMSample/WpfMVVMSample/Converters/EnumBindingSource.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+            Type actualEnumType = Nullable.GetUnderlyingType(this._enumType) ?? this._enumType;
+            Array enumValues = Enum.GetValues(actualEnumType);
+
+            //if (actualEnumType == this._enumType)
+            //    return enumValues;
+
+            //ommits the first enum element, typically "undefined"
+            Array tempArray = Array.CreateInstance(actualEnumType, enumValues.Length -1);
+            Array.ConstrainedCopy(enumValues,1,tempArray, 0,enumValues.Length-1);
```
[the remark this answers is not in the document]
Comment (dayewah) [reply Samples/WpfMVVMSample/WpfMVVMSample/Converters/EnumBindingSource.cs]: Great tip. Thought I was stuck with the old array methods. Just needed a simple cast<object>() to simplify things.

File: Samples/WpfMVVMSample/WpfMVVMSample/Converters/UnitToStringConverter.cs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
+        {
+            var unitType = value.GetType();
+            var unitEnumType = unitType.GetProperty("BaseUnit").PropertyType;
+            var unitEnumValue = _settings.GetDefaultUnit(unitEnumType);
+            var significantDigits = _settings.SignificantDigits;
+
+            var result = unitType
+                    .GetMethod("ToString", new[] { unitEnumType, typeof(IFormatProvider), typeof(int) })
+                    .Invoke(value, new object[] { unitEnumValue, null, significantDigits });
```
[the remark this answers is not in the document]
Comment (dayewah) [reply Samples/WpfMVVMSample/WpfMVVMSample/Converters/UnitToStringConverter.cs]: I agree that the reflection is a bit clunky. Any library enhancements that better enable this strategy would be welcome.

File: Samples/WpfMVVMSample/WpfMVVMSample/Converters/EnumBindingSource.cs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+        public override object ProvideValue(IServiceProvider serviceProvider)
+        {
+            if (this._enumType== null)
+                throw new InvalidOperationException("The EnumType must be specified.");
+
+            Type actualEnumType = Nullable.GetUnderlyingType(this._enumType) ?? this._enumType;
+            Array enumValues = Enum.GetValues(actualEnumType);
+
+            //if (actualEnumType == this._enumType)
```
[the remark this answers is not in the document]
Comment (dayewah) [reply Samples/WpfMVVMSample/WpfMVVMSample/Converters/EnumBindingSource.cs]: Will do.

**Comment by dayewah** [comment]
Updates complete

File: Samples/WpfMVVMSample/WpfMVVMSample/Converters/UnitToStringConverter.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ferittuncer) [reply Samples/WpfMVVMSample/WpfMVVMSample/Converters/UnitToStringConverter.cs]: Well, we have experienced these problems too when we used this library in our GUI. Lacking base classes for units and quantities forces implementation to reflection as presenter layer doesn't know which unit or quantity it's dealing with at the runtime.

Perhaps there are other ways we couldn't see, I'm not sure. I just have an instinct that we should have base classes. You know better, your call @angularsen

[the remark this answers is not in the document]
Comment (eriove) [reply Samples/WpfMVVMSample/WpfMVVMSample/Converters/UnitToStringConverter.cs]: A common interface for all quantities would be a simple addition (at least if it contained members that are already there) and wouldn't incur any performance hit (boxing) as long as it isn't used.

That's of course not part of this pull request but it is much easier to see in this context.

**Comment by angularsen** [comment]
Alright, so I think we agree this is ready to merge and that there is some possibility to improve this sample by looking for ways for the library to provide the information that is now gathered by the reflection code.

**Comment by angularsen** [comment]
Thanks @dayewah !

**Comment by angularsen** [comment]
I won't personally pursue the case of improving the reflection code as discussed above, but at least this gives us a concrete usecase to refer to in future discussion. If someone wants to draft up a proposal on how to avoid reflection code in this sample and similar usecases, just create an issue for it and I would be happy to participate in the discussion and bring that into the library.

**Comment by dayewah** [comment]
Great! Thanks @angularsen. I'll continue to monitor progress with library changes and perhaps make suggestions if I come up with anything.

**Pull request merged**